dtfs-central-api/src/services/state-machines/utilisation-report/event-handlers/generate-keying-data/generate-keying-data.event-handler.ts

```diff
@@ -1,29 +1,55 @@
-import { EntityManager } from 'typeorm';
-import { DbRequestSource, UtilisationReportEntity } from '@ukef/dtfs2-common';
+import { EntityManager, In } from 'typeorm';
+import { DbRequestSource, FeeRecordEntity, FeeRecordStatus, UtilisationReportEntity } from '@ukef/dtfs2-common';
 import { BaseUtilisationReportEvent } from '../../event/base-utilisation-report.event';
-import { GenerateKeyingDataDetails } from '../../../../../v1/controllers/utilisation-report-service/post-keying-data.controller/helpers';
 import { FeeRecordStateMachine } from '../../../fee-record/fee-record.state-machine';
 
+const getFacilityIdsWhereKeyingDataCannotBeCalculated = async (entityManager: EntityManager, reportId: number): Promise<Set<string>> => {
+  const feeRecordsWhereKeyingDataCannotBeCalculated = await entityManager.find(FeeRecordEntity, {
+    where: {
+      report: { id: reportId },
+      status: In<FeeRecordStatus>(['TO_DO', 'DOES_NOT_MATCH']),
+    },
+  });
+  return feeRecordsWhereKeyingDataCannotBeCalculated.reduce((facilityIds, { facilityId }) => facilityIds.add(facilityId), new Set<string>());
+};
+
 type GenerateKeyingDataEventPayload = {
   transactionEntityManager: EntityManager;
-  generateKeyingDataDetails: GenerateKeyingDataDetails;
+  matchFeeRecords: FeeRecordEntity[];
   requestSource: DbRequestSource;
 };
 
 export type UtilisationReportGenerateKeyingDataEvent = BaseUtilisationReportEvent<'GENERATE_KEYING_DATA', GenerateKeyingDataEventPayload>;
 
 export const handleUtilisationReportGenerateKeyingDataEvent = async (
   report: UtilisationReportEntity,
-  { transactionEntityManager, generateKeyingDataDetails, requestSource }: GenerateKeyingDataEventPayload,
+  { transactionEntityManager, matchFeeRecords, requestSource }: GenerateKeyingDataEventPayload,
 ): Promise<UtilisationReportEntity> => {
+  const facilityIdsWhereKeyingDataCannotBeCalculated = await getFacilityIdsWhereKeyingDataCannotBeCalculated(transactionEntityManager, report.id);
+  const processedFacilityIds = new Set<string>();
+
   await Promise.all(
-    generateKeyingDataDetails.map(({ feeRecord, generateKeyingData }) => {
+    matchFeeRecords.map((feeRecord) => {
+      const { facilityId } = feeRecord;
       const stateMachine = FeeRecordStateMachine.forFeeRecord(feeRecord);
+
+      if (processedFacilityIds.has(facilityId)) {
+        return stateMachine.handleEvent({
+          type: 'GENERATE_KEYING_DATA',
+          payload: {
+            transactionEntityManager,
+            isFinalFeeRecordForFacility: false,
+            requestSource,
+          },
+        });
+      }
```

Comment:
I think this would be slightly cleaner if you kept the mapping fee records to objects pairing them with the isFinalFeeRecordForFacility and then mapped through that to call the fee record state machine. It is another loop through, but I think it would be more readable, and I think having a type for that pairing is less opaque and confusing when it would just be in this file and not exported? I'm not too set on this though so I'll leave it up to you.

Reply:
Sure - I try to avoid looping through twice, but if you think it makes it clearer then I don't mind